File: packages/shared/src/agent/tools/quality.js
'use strict';

const fs = require('fs');
const path = require('path');
const { execTool, resolveWorkdir, commandExists, platformCmd, defineTool } = require('./exec');

/**
 * Quality tools — lint/format and run tests by delegating to the project's own
 * linters and test runners (ESLint, Prettier, Ruff, Black, npm test, pytest,
 * Gradle, Go, Cargo, Maven). Detection is by project files; no runner logic is
 * re-implemented.
 */

const anyFile = (dir, names) => names.some((n) => fs.existsSync(path.join(dir, n)));

/** Choose a linter/formatter for a directory. */
async function pickLinter(dir, mode, requested) {
  const fix = mode === 'fix';
  const linters = {
    eslint: { command: platformCmd('npx'), args: ['--no-install', 'eslint', '.', ...(fix ? ['--fix'] : [])], hint: 'Add ESLint to the project (npm i -D eslint).' },
    prettier: { command: platformCmd('npx'), args: ['--no-install', 'prettier', fix ? '--write' : '--check', '.'], hint: 'Add Prettier (npm i -D prettier).' },
    ruff: { command: 'ruff', args: fix ? ['check', '--fix', '.'] : ['check', '.'], hint: 'pip install ruff.' },
    black: { command: 'black', args: fix ? ['.'] : ['--check', '.'], hint: 'pip install black.' },
  };
  if (requested && requested !== 'auto') return { key: requested, ...linters[requested] };
  if (anyFile(dir, ['.eslintrc', '.eslintrc.js', '.eslintrc.json', '.eslintrc.cjs', 'eslint.config.js', 'eslint.config.mjs'])) return { key: 'eslint', ...linters.eslint };
  if (anyFile(dir, ['.prettierrc', '.prettierrc.json', '.prettierrc.js', 'prettier.config.js']) || fs.existsSync(path.join(dir, 'package.json'))) return { key: 'prettier', ...linters.prettier };
  if (anyFile(dir, ['pyproject.toml', 'ruff.toml', '.ruff.toml']) && (await commandExists('ruff'))) return { key: 'ruff', ...linters.ruff };
  if (anyFile(dir, ['pyproject.toml', 'setup.py', 'requirements.txt'])) return { key: 'black', ...linters.black };
  return { key: 'prettier', ...linters.prettier };
}

/** Choose a test runner (pure). */
function pickTestRunner(dir, requested) {
  const has = (f) => fs.existsSync(path.join(dir, f));
  const readPkg = () => {
    try {
      return JSON.parse(fs.readFileSync(path.join(dir, 'package.json'), 'utf8'));
    } catch (_) {
      return null;
    }
  };
  const runners = {
    npm: () => ({ command: platformCmd('npm'), args: ['test'], hint: 'Install Node.js (bundles npm).' }),
    pytest: () => ({ command: 'pytest', args: ['-q'], hint: 'pip install pytest.' }),
    gradle: () => ({ command: process.platform === 'win32' ? 'gradlew.bat' : './gradlew', args: ['test'], hint: 'Use the Gradle wrapper (gradlew).' }),
    maven: () => ({ command: has('mvnw') ? (process.platform === 'win32' ? 'mvnw.cmd' : './mvnw') : 'mvn', args: ['-q', 'test'], hint: 'Use the Maven wrapper (mvnw) or install Maven.' }),
    go: () => ({ command: 'go', args: ['test', './...'], hint: 'Install Go.' }),
    cargo: () => ({ command: 'cargo', args: ['test'], hint: 'Install the Rust toolchain (rustup).' }),
  };
  if (requested && requested !== 'auto' && runners[requested]) return { key: requested, ...runners[requested]() };
  const pkg = readPkg();
  if (pkg && pkg.scripts && pkg.scripts.test) return { key: 'npm', ...runners.npm() };
  if (has('pytest.ini') || has('tox.ini') || has('pyproject.toml') || has('setup.py')) return { key: 'pytest', ...runners.pytest() };
  if (has('gradlew') || has('build.gradle') || has('build.gradle.kts')) return { key: 'gradle', ...runners.gradle() };
  if (has('pom.xml')) return { key: 'maven', ...runners.maven() };
  if (has('go.mod')) return { key: 'go', ...runners.go() };
  if (has('Cargo.toml')) return { key: 'cargo', ...runners.cargo() };
  return null;
}

const lintFormatTool = defineTool(
  {
    name: 'lint_format',
    description:
      'Lint and/or format the workspace using the project\'s configured tools (ESLint, Prettier, Ruff, Black). ' +
      'mode "check" (default) reports issues; mode "fix" applies safe autofixes. Prefer this over manual formatting.',
    schema: (z) =>
      z.object({
        dir: z.string().optional().describe('workspace-relative directory'),
        mode: z.enum(['check', 'fix']).optional().describe('default: check'),
        tool: z.enum(['auto', 'eslint', 'prettier', 'ruff', 'black']).optional().describe('force a tool (default: auto)'),
      }),
  },
  async (input, ctx) => {
    const dir = resolveWorkdir(ctx, input.dir);
    const l = await pickLinter(dir, input.mode || 'check', input.tool);
    return execTool({ ctx, label: `lint/format (${l.key}, ${input.mode || 'check'})`, command: l.command, args: l.args, dir: input.dir, notFoundHint: l.hint });
  }
);

const testRunTool = defineTool(
  {
    name: 'test_run',
    description:
      'Run the project test suite using its native runner (npm test, pytest, Gradle, Maven, Go, Cargo), ' +
      'auto-detected from the workspace. Prefer this over guessing the test command.',
    schema: (z) =>
      z.object({
        dir: z.string().optional().describe('workspace-relative directory'),
        system: z.enum(['auto', 'npm', 'pytest', 'gradle', 'maven', 'go', 'cargo']).optional().describe('force a runner (default: auto)'),
      }),
  },
  async (input, ctx) => {
    const dir = resolveWorkdir(ctx, input.dir);
    const r = pickTestRunner(dir, input.system);
    if (!r) return `No known test runner detected in ${dir} (looked for npm test / pytest / Gradle / Maven / Go / Cargo).`;
    return execTool({ ctx, label: `tests (${r.key})`, command: r.command, args: r.args, dir: input.dir, notFoundHint: r.hint });
  }
);

const FACTORIES = Object.freeze({ lint_format: lintFormatTool, test_run: testRunTool });

module.exports = { FACTORIES, pickLinter, pickTestRunner };

File: ai_fleet/agent/tools/quality.py
"""Quality tools (port of agent/tools/quality.js).

Lint/format and run tests by delegating to the project's own linters and test
runners (ESLint, Prettier, Ruff, Black, npm test, pytest, Gradle, Go, Cargo,
Maven). Detection is by project files; no runner logic is re-implemented.
"""

from __future__ import annotations

import json
import os
import sys
from typing import Literal, Optional

from pydantic import BaseModel, Field

from ai_fleet.agent.tools.exec import (
    command_exists,
    define_tool,
    exec_tool,
    platform_cmd,
    resolve_workdir,
)


def _any_file(dir, names):
    return any(os.path.exists(os.path.join(dir, n)) for n in names)


async def pick_linter(dir, mode, requested=None):
    """Choose a linter/formatter for a directory."""
    fix = mode == "fix"
    linters = {
        "eslint": {
            "command": platform_cmd("npx"),
            "args": ["--no-install", "eslint", ".", *(["--fix"] if fix else [])],
            "hint": "Add ESLint to the project (npm i -D eslint).",
        },
        "prettier": {
            "command": platform_cmd("npx"),
            "args": ["--no-install", "prettier", "--write" if fix else "--check", "."],
            "hint": "Add Prettier (npm i -D prettier).",
        },
        "ruff": {
            "command": "ruff",
            "args": ["check", "--fix", "."] if fix else ["check", "."],
            "hint": "pip install ruff.",
        },
        "black": {
            "command": "black",
            "args": ["."] if fix else ["--check", "."],
            "hint": "pip install black.",
        },
    }
    if requested and requested != "auto":
        return {"key": requested, **linters[requested]}
    if _any_file(dir, [".eslintrc", ".eslintrc.js", ".eslintrc.json", ".eslintrc.cjs", "eslint.config.js", "eslint.config.mjs"]):
        return {"key": "eslint", **linters["eslint"]}
    if _any_file(dir, [".prettierrc", ".prettierrc.json", ".prettierrc.js", "prettier.config.js"]) or os.path.exists(
        os.path.join(dir, "package.json")
    ):
        return {"key": "prettier", **linters["prettier"]}
    if _any_file(dir, ["pyproject.toml", "ruff.toml", ".ruff.toml"]) and (await command_exists("ruff")):
        return {"key": "ruff", **linters["ruff"]}
    if _any_file(dir, ["pyproject.toml", "setup.py", "requirements.txt"]):
        return {"key": "black", **linters["black"]}
    return {"key": "prettier", **linters["prettier"]}


def pick_test_runner(dir, requested=None):
    """Choose a test runner (pure)."""
    def has(f):
        return os.path.exists(os.path.join(dir, f))

    def read_pkg():
        try:
            with open(os.path.join(dir, "package.json"), "r", encoding="utf-8") as f:
                return json.load(f)
        except Exception:
            return None

    runners = {
        "npm": lambda: {"command": platform_cmd("npm"), "args": ["test"], "hint": "Install Node.js (bundles npm)."},
        "pytest": lambda: {"command": "pytest", "args": ["-q"], "hint": "pip install pytest."},
        "gradle": lambda: {
            "command": "gradlew.bat" if sys.platform.startswith("win") else "./gradlew",
            "args": ["test"],
            "hint": "Use the Gradle wrapper (gradlew).",
        },
        "maven": lambda: {
            "command": (("mvnw.cmd" if sys.platform.startswith("win") else "./mvnw") if has("mvnw") else "mvn"),
            "args": ["-q", "test"],
            "hint": "Use the Maven wrapper (mvnw) or install Maven.",
        },
        "go": lambda: {"command": "go", "args": ["test", "./..."], "hint": "Install Go."},
        "cargo": lambda: {"command": "cargo", "args": ["test"], "hint": "Install the Rust toolchain (rustup)."},
    }
    if requested and requested != "auto" and requested in runners:
        return {"key": requested, **runners[requested]()}
    pkg = read_pkg()
    if pkg and pkg.get("scripts") and pkg["scripts"].get("test"):
        return {"key": "npm", **runners["npm"]()}
    if has("pytest.ini") or has("tox.ini") or has("pyproject.toml") or has("setup.py"):
        return {"key": "pytest", **runners["pytest"]()}
    if has("gradlew") or has("build.gradle") or has("build.gradle.kts"):
        return {"key": "gradle", **runners["gradle"]()}
    if has("pom.xml"):
        return {"key": "maven", **runners["maven"]()}
    if has("go.mod"):
        return {"key": "go", **runners["go"]()}
    if has("Cargo.toml"):
        return {"key": "cargo", **runners["cargo"]()}
    return None


class _LintFormatSchema(BaseModel):
    dir: Optional[str] = Field(default=None, description="workspace-relative directory")
    mode: Optional[Literal["check", "fix"]] = Field(default=None, description="default: check")
    tool: Optional[Literal["auto", "eslint", "prettier", "ruff", "black"]] = Field(
        default=None, description="force a tool (default: auto)"
    )


class _TestRunSchema(BaseModel):
    dir: Optional[str] = Field(default=None, description="workspace-relative directory")
    system: Optional[Literal["auto", "npm", "pytest", "gradle", "maven", "go", "cargo"]] = Field(
        default=None, description="force a runner (default: auto)"
    )


async def _lint_format(input, ctx):
    dir_ = resolve_workdir(ctx, input.get("dir"))
    mode = input.get("mode") or "check"
    linter = await pick_linter(dir_, mode, input.get("tool"))
    return await exec_tool(
        ctx=ctx,
        label=f"lint/format ({linter['key']}, {mode})",
        command=linter["command"],
        args=linter["args"],
        dir=input.get("dir"),
        not_found_hint=linter["hint"],
    )


async def _test_run(input, ctx):
    dir_ = resolve_workdir(ctx, input.get("dir"))
    runner = pick_test_runner(dir_, input.get("system"))
    if not runner:
        return f"No known test runner detected in {dir_} (looked for npm test / pytest / Gradle / Maven / Go / Cargo)."
    return await exec_tool(
        ctx=ctx,
        label=f"tests ({runner['key']})",
        command=runner["command"],
        args=runner["args"],
        dir=input.get("dir"),
        not_found_hint=runner["hint"],
    )


lint_format_tool = define_tool(
    {
        "name": "lint_format",
        "description": (
            "Lint and/or format the workspace using the project's configured tools (ESLint, Prettier, Ruff, Black). "
            'mode "check" (default) reports issues; mode "fix" applies safe autofixes. Prefer this over manual formatting.'
        ),
        "schema": _LintFormatSchema,
    },
    _lint_format,
)

test_run_tool = define_tool(
    {
        "name": "test_run",
        "description": (
            "Run the project test suite using its native runner (npm test, pytest, Gradle, Maven, Go, Cargo), "
            "auto-detected from the workspace. Prefer this over guessing the test command."
        ),
        "schema": _TestRunSchema,
    },
    _test_run,
)

FACTORIES = {"lint_format": lint_format_tool, "test_run": test_run_tool}
